### PR TITLE
Add array read and update function to SMT2 backend

### PR DESCRIPTION
**Pull request by esteffin**
Add array read and write function to the SMT2 backend.
Specifically:
 - `index_exprt` is translated to SMT2 `select` function,
 - `with_exprt` is translated to SMT2 `store` function.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: unit/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp [hunk 1]
```diff
@@ -170,7 +170,7 @@ TEST_CASE(
       std::vector<smt_commandt>{
         smt_set_option_commandt{smt_option_produce_modelst{true}},
         smt_set_logic_commandt{
-          smt_logic_quantifier_free_uninterpreted_functions_bit_vectorst{}},
+          smt_logic_quantifier_free_uninterpreted_functions_bit_vectors_arrayst{}},
```
Comment (TGWDB) [inline unit/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp]: Unimportant, but should the order of names match the order of letters?

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 2]
```diff
@@ -96,6 +96,20 @@ expr_to_smt_conversion_test_environmentt::convert(const exprt &expression) const
...
+
+  const auto array_type =
+    array_typet{signedbv_typet{8}, from_integer(8, c_index_type())};
+  INFO("Type being converted: " + array_type.pretty(2, 0));
+  const auto conversion_result = convert_type_to_smt_sort(array_type);
+  CHECK(
+    conversion_result ==
+    smt_array_sortt{smt_bit_vector_sortt{64}, smt_bit_vector_sortt{8}});
+}
+
```
Comment (TGWDB) [inline unit/solvers/smt2_incremental/convert_expr_to_smt.cpp:113]: I think the spec allows for multiple sorts as both index and value sorts? We don't test many here.

File: regression/cbmc-incr-smt2/arrays/array_read.desc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+CORE
+array_read.c
+
+Passing problem to incremental SMT2 solving
```
Comment (NlightNFotis) [inline regression/cbmc-incr-smt2/arrays/array_read.desc:4]: Can we also get a regex that checks for the condition failure?

It kind of implicitly does because of the check for the `EXIT=10`, but it would be good to have one checking the `__CPROVER_assert` fails.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 4]
```diff
@@ -991,10 +991,27 @@ static smt_termt convert_expr_to_smt(
...
+  return smt_array_theoryt::store(old_array_term, index_term, value_term);
+}
+
 static smt_termt convert_expr_to_smt(
   const with_exprt &with,
   const sub_expression_mapt &converted)
 {
+  if(const auto array_type = type_try_dynamic_cast<array_typet>(with.type()))
+  {
+    return convert_array_update_to_smt(
```
Comment (NlightNFotis) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:1012]: Question: why are we dispatching to a new function that handles only this particular case? Doesn't it make more sense to inline `convert_array_update_to_smt` within `convert_expr_to_smt`?

Unless you expect `convert_array_update_to_smt` to be reused.

File: regression/cbmc-incr-smt2/arrays/array_read.desc (hunk 3 above)
Comment (TGWDB) [reply regression/cbmc-incr-smt2/arrays/array_read.desc:4]: Also exit signal may be OS dependant?

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 4 above)
Comment (NlightNFotis) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp:1012]: @esteffin Can we add a comment that `with_exprt`s may also be handling structs, so this dispatches to that function to handle arrays?

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 2 above)
Comment (thomasspriggs) [reply unit/solvers/smt2_incremental/convert_expr_to_smt.cpp:113]: The question is which spec? The specification for ArraysEx allows sorts of Arrays of any sort of index and value. However we are targeting the logic `QF_AUFBV`. This logic only supports arrays with bit vector index and element sorts. Therefore this PR is currently missing the piece of work for converting booleans to bit vectors in the case where we have an cbmc array of booleans.

TLDR We should be putting arrays of nonbitvector type into the conversions and checking that we always get arrays of bit vectors out.

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 5]
```diff
@@ -1140,6 +1141,30 @@ TEST_CASE(
   }
 }
 
+TEST_CASE(
+  "expr to smt conversion for index_exprt expressions",
+  "[core][smt2_incremental]")
+{
+  auto test =
+    expr_to_smt_conversion_test_environmentt::make(test_archt::x86_64);
+  SECTION("Array types")
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :pick: Given that `index_exprt` is applicable only to array types, I don't think it is worth having an "Array types" section. It is worthwhile for `with_exprt`, because `with_exprt` can also be used for structs.

File: regression/cbmc-incr-smt2/arrays/array_read.desc (hunk 3 above)
Comment (thomasspriggs) [reply regression/cbmc-incr-smt2/arrays/array_read.desc:4]: We can check just the condition failure by matching the string "line 6 Array condition: FAILURE"

Comment (esteffin) [reply regression/cbmc-incr-smt2/arrays/array_read.desc:4]: Changed

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 4 above)
Comment (esteffin) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp:1012]: Comment added

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 2 above)
Comment (esteffin) [reply unit/solvers/smt2_incremental/convert_expr_to_smt.cpp:113]: With the current changes CBMC will allow multi-dimensional array and boolean array support.
The SMT backend (z3 and cvc5) support that, except z3 on ubuntu 18.04 (z3 version 4.4.0) that rejects it as it is not SMTlib 2.6.

My suggestion is to keep this as is and without additional tests (that would fail on ubuntu 18.04).

**Review by thomasspriggs — APPROVED** [review]

**Review by NlightNFotis — APPROVED** [review]
